metadata/src/utils/retain.rs

```diff
@@ -261,13 +404,62 @@ pub fn retain_metadata<F, G>(
         .iter()
         .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
         .expect("Metadata must contain sp_runtime::DispatchError");
-    type_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(metadata.runtime_ty);
+    type_set.collect_extrinsic_types(&metadata.extrinsic);
+
+    // Collect the outer enums type IDs.
+    for typ in [
+        metadata.outer_enums.call_enum_ty,
+        metadata.outer_enums.event_enum_ty,
+        metadata.outer_enums.error_enum_ty,
+    ] {
+        type_set.collect_variants_in_type(metadata, typ, &mut pallets_filter);
+    }
 
-    // Now, keep the type IDs we've asked for. This recursively keeps any types referenced from these.
-    // This will return a map from old to new type ID, because IDs may change.
-    let map_ids = metadata.types.retain(|id| type_ids.contains(&id));
+    let mut retained_set = TypeSet {
+        seen_ids: type_set
+            .seen_ids
+            .difference(&retained_set.seen_ids)
+            .copied()
+            .collect(),
+    };
```

Comment:
So it looks like you are:
- building a set of types to keep based on the pallets we want to keep (`retained_set`)
- cloning that set and then adding all of the other types we want to keep (runtime APIs etc) too (`type_set`)
- taking the difference between those (ie `type_set - retained_set`? So just the other types without the pallet ones, which we could get by just not cloning above?). This difference is now `retained_set`.
- for pallets we want to filter, run `update_filtered_pallet`, which roughly seems to keep things in the pallet if they reference types we've collected outside of the pallet stuff, and remove things if not. We then only discard pallets if no types that they directly reference are used elsewhere outside of pallets.

I might be wrong, but I don't really understand the reason for doing this? since you collected the outer enums, I can see that the first pallet we see that references those would actually be kept (which I guess has the effect of keeping the variant too?), but it feels like quite a long winded way to go about avoiding getting rid of variants if the outer enums are used somewhere, and I don't know if there is any other benefit :D

I also see that `update_filtered_pallet` calls `retained_set.remove()` each time it checks, so only the first pallet referencing eg RuntimeCall would end up having bits kept I think, and other pallets that reference `RuntimeCall` would be removed. Is that intentional?

Overall, I feel like we can remove a lot of this logic and rely on `scale_info::PortableRegistry::retain` for the recursive retaining of IDs we care about (because either that works ok already, or doesn't work and should be fixed). Then the logic can be something like:
- Add all of the type IDs from the pallets etc that we want to keep to a set
- For any return types or storage types in these things we are keeping, do a deep scan to see if we find the `RuntimeCall`, `RutimeEvent` or `RuntimeError` outer enums. If we find one, we can't strip any variants out of it
- Now, call `PortableRegistry::retain` to keep all of the type Ids we need (which will vary depending on whether those enums had variants stripped out of them or not and thus fewer IDs to recurse through).

All of that said though, if there is some value I'm not understanding in the way that we are keeping pallets etc here, then probably I'd need to be walked through it and we'd want to add a comment or two somewhere here to explain the approach and reasoning for it for our future selves :)

Reply:
> taking the difference between those (ie type_set - retained_set? So just the other types without the pallet ones, which we could get by just not cloning above?). This difference is now retained_set. 

The difference is the types present in the `type_set` and absent in `retained_set`. `retained_set` is not interesting in this case as types were already present in retained pallet. But i feel like we are talking about the same thing 

> but it feels like quite a long winded way to go about avoiding getting rid of variants if the outer enums are used somewhere, and I don't know if there is any other benefit :D

Its more about getting rid of pallets and outer enums, *but only if they are not referenced anywhere*. Its about stripping types that are not interesting to user without accidentally removing something that is required.

> I also see that update_filtered_pallet calls retained_set.remove() each time it checks, so only the first pallet referencing eg RuntimeCall would end up having bits kept I think, and other pallets that reference RuntimeCall would be removed. Is that intentional?

I guess i'm misunderstanding something, shouldn't only one pallet with the type be enough or we can safely ignore those refs to `runtime call`/`runtime error`/`runtime event` and remove pallet based purely on storage/constant types? My intuition was to keep them as they might be referencing inner types inside outerenum variant, ie `pallet.event_ty` = 365 and `metadata.outer_enum.event_ty = 22`